tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc

```diff
@@ -131,6 +132,8 @@ Status NVPTXCompiler::OptimizeHloConvolutionCanonicalization(
                                           /*allow_mixed_precision=*/false);
 
     AlgebraicSimplifierOptions options;
+    options.set_cudnn_batchnorm_forward_training_metadata(
+        kCudnnBatchNormForwardTrainingCallTarget);
```

Comment:
This custom call is created by Visitor::HandleBatchNormTraining. Can we run algebraic_simplifier before that and use the fact "batchnorm HLO returns {output, mean, variance} and variance is non-negative" to perform the optimization? In this case, I think you may not even need to rely on AlgebraicSimplifierOptions turn on/off the optimization.

Reply:
I have thought about this but I couldn't not come to the conclusion that the pattern (motivation pattern mentioned in the description) emerges before the rewriter pass. What I can do is add another case to `IsNonNegative` (`IsPositive`) that handles batchnorm HLO. That seems to be a natural extension anyway.

Reply:
Can you provide some detail on how the original code and the code generated by Visitor::HandleBatchNormTraining interact and produce the pattern? What I can tell is the transformation replaces a use of the x (batch norm result with index 2) with this function of y (custom call result with index 2):  power(y, -2) - epsilon.

Reply:
Before `cudnn_batchnorm_rewiter`, the pattern that exist is `Inv([rsqrt(x+c)*rsqrt(x+c)])-c` which gets transformed to the one handled in this PR. This can also be simplified to `x` but might prove to be non-trivial. I can check if this pattern gets eliminated by existing transforms+the ones I've added here. However, do you think we need to add this case as well?